Support conditional requirement for -c argument

begriffs opened the pull request:
Fixes #736

The `readOptions` function is getting ungainly. I added comments to help explain it. The notion of having a command line argument that is conditionally mandatory does not fit Haskell's notion of Applicative, hence using optparse-applicative to parse the args is awkward. However that library is pretty convenient with help text formatting etc that I kept it and am working around it a little.

Code reviews appreciated.

ruslantalpa commented:
Looking at the example from https://github.com/pcapriotti/optparse-applicative
it seems you can have a mandatory config but also a help switch.
Why not have this behaviour (which seems to be supported by opt parse)
- no arguments -- display help
- help argument -- display help
- -c argument -- normal usage

help should show only info for 2 arguments -c and -h (help) as parameters for for command lines and if you want also display a sample config when help is invoke rather then having a special parameter to extract info about a sample config, include it in the help screen

ruslantalpa commented:
`progDesc` could be used for the sample config file

begriffs commented:
Good point, the help text doesn't have much to say so might as well output the example config file too. I changed it to include the example at the end of the message (as a footer). I removed the `-c` option. There is only one argument now, just the config filename. All other flags can go into the config file, keeping the command line simple.

One nice property of the old way was that `postgrest --example-config > pg.conf` outputs exactly the lines necessary for the config file whereas in this new way the user has to split and dedent them from the usage message. Not too big of a deal though.

My final final thought is that the usage text might get super long if the config file gets a lot of new options. At that point we could choose to output a minimal config with a link to full documentation as a comment.
